packages/data/src/redux-store/index.js

```diff
@@ -598,6 +598,9 @@ function mapSelectorsWithResolvers(
 		}
 
 		const selectorResolver = ( ...args ) => {
+			if ( resolver.normalizeArgs ) {
```

Comment:
We'd need to check if it's a callable function...

Reply:
This code currently passes the normalized args also to the `selector` itself, but I think we want to pass them only to the `fulfillSelector` function.

@youknowriad what do you think about the new `normalizeArgs` optional resolver field? The problem being solved is that the following selector calls are equivalent:
```js
getEntityRecord( 'postType', 'post', 1 );
getEntityRecord( 'postType', 'post', '1' );
```
But the resolver machinery (`resolversCache`, metadata reducer) will treat these two `args` arrays as different, and will store two distinct records in the `EquivalentKeyMap`s that hold the resolution state.

At the same time we are reluctant to do the string->number conversion in the `selectorArgsToStateKey` function because this normalization is application-specific, other selectors might not want it.

Reply:
...and bear in mind that we cannot simply normalize the args in the selector or resolver function body itself because the `args` are used by the aforesaid "resolver machinery" prior to invocation of the functions themselves.

Reply:
> This code currently passes the normalized args also to the selector itself, but I think we want to pass them only to the fulfillSelector function.

I would think we'd like to normalize the arguments for _both_ selector and resolver? I would expect both to receive the same arguments.

------

Another option we could consider is having a property like `normalizeArgsToCacheKey` (again on the resolver) which we can use selectively _only_ at the points where we interact with the cache thereby leaving the `args` "intact" elsewhere.

Reply:
There are two ways how to work with the normalization:
1. limit it only to the places that interact with the internal resolvers cache and resolution metadata state. The resolver and the selector itself would get the raw args. It's user-space code, it's supposed to handle them on its own.
2. Always normalize them and pass the normalized version to the selector and the resolvers, too.

I think both ways are good, but you're right that 2) is probably better. The user-space selectors and resolvers likely want to do the same normalization anyway.

In addition to converting strings to numbers, another good use case for normalization is supplying default values. Currently, a selector like:
```js
function getBlocks( type, filter = 'all' ) {}
```
will store two non-equivalent args keys in the resolvers cache: `[ type, undefined ]` and `[ type, 'all' ]`. Normalization can merge them into one.

Reply:
Agreed here, I think if we normalize, we should normalize for all places (for selectors that define the normalize function). Which makes me thing maybe the normalization is a property of a selector (and not the resolver) since not all selectors have resolvers.

Reply:
OK I'll move this to the selector 👍

Reply:
✅